### PR TITLE
fixed race where build start event could be fired multiple times.

### PR DESCRIPTION
**Pull request by heejaechang**
fix https://github.com/dotnet/roslyn/issues/15047

### CODE REVIEW

**Comment by heejaechang** [comment]
@mavasani @srivatsn @panopticoncentral can you take a look?

File: src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs [hunk 1]
```diff
@@ -153,10 +156,10 @@ internal void OnSolutionBuild(object sender, UIContextChangedEventArgs e)
             }
 
             // get local copy of inprogress state
-            var inprogressState = _state;
+            var inprogressState = BuildInprogressState;
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs]: read + write should proably both e done under a lock.

File: src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs [hunk 2]
```diff
@@ -342,17 +345,42 @@ public void AddNewErrors(DocumentId documentId, DiagnosticData diagnostic)
...
+            lock (_gate)
+            {
+                _stateDoNotAccessDirectly = null;
+            }
+        }
+
         private InprogressState GetOrCreateInprogressState()
         {
-            if (_state == null)
+            if (_stateDoNotAccessDirectly == null)
```
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs]: why is it not safe? the issue was state might created multiple time even if it always return right one. this should never make state to be created multiple time in same session.

File: src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs (hunk 1 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs]: sure.

**Comment by CyrusNajmabadi** [comment]
> why is it not safe? the issue was state might created multiple time even if it always return right one. this should never make state to be created multiple time in same session

Because callers can observe reordered writes.

i.e. they may see the write to _stateDoNotAccessDirectly before the constructor call of "new InprogressState(this, _workspace.CurrentSolution)" has actually completed.

Writes can be reordered, and if you're not properly synchronizing, that's observable and unsafe.

**Comment by CyrusNajmabadi** [comment]
See http://csharpindepth.com/Articles/General/Singleton.aspx#dcl

**Comment by heejaechang** [comment]
ping?

**Comment by CyrusNajmabadi** [comment]
👍

**Pull request merged**